### PR TITLE
Fix #1524 and other softmax mask functionality

### LINKED ISSUE #1524: Causal mask ignored in DotProductAttention
for transformer_engine version 1.14.0+87fbe812f
in transformer_engine.jax.flax.module
in Softmax class

in __call__ line 191-198:

```
# For the case that self.softmax == SoftmaxType.SCALED_UPPER_TRIANG_MASKED
# and kernel is unavailable, then try on pure scaled softmax custom calls.
if is_softmax_kernel_available(
    SoftmaxType.SCALED, batch, heads, q_seqlen, k_seqlen, dtype
):
    outputs = softmax(logits, None, self.scale_factor, SoftmaxType.SCALED)
else:
    outputs = jax_nn.softmax(logits * self.scale_factor)  # <- self.softmax_type ignored
```

after else, self.softmax_type is ignored and no causal attention is performed when e.g. self.softmax_type = SoftmaxType.SCALED_UPPER_TRIANG_MASKED

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

Fixes #1524 
Simplifies and corrects logic around mapping attn mask to softmax mask, while adding additional testing capabilities

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

Please list the changes introduced in this PR:

1. Add test cases for better coverage in `test_layer.py`

- causal and window size None
- causal and window size default (-1,1)
- no_mask and window size default (-1,1)
- no_mask and window size default (2,2)
- padding and window size None
- padding_causal and window_size (2,2)

2. Correctly map `padding_causal` to `scaled_masked` instead of `scaled_upper_triangle`
3. Add fallback jax implementation for `scaled_upper_triangle` (#1524 )
4. Add tests for `Softmax` wrapper/module in `jax.flax.module.Softmax`
5. Simplify logic for jax based softmax
6. Add support table in comments for attn mask, SWA and softmax type

## TODO:
Add support for `causal_bottom_right` and `padding_causal_bottom_right` as well. 

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_layer.py [hunk 1]
```diff
@@ -215,12 +215,53 @@ def enable_fused_attn():
...
+        _KEY_OF_RELATIVE_EMBEDDING: False,
+        _KEY_OF_SELF_ATTN_MASK_TYPE: "padding",
+        _KEY_OF_WINDOW_SIZE: None,
+    },
+    # attrs28
+    {
+        _KEY_OF_TRANSPOSE_BS: False,
+        _KEY_OF_RELATIVE_EMBEDDING: False,
+        _KEY_OF_WINDOW_SIZE: (2, 2),
+    },
```
Comment (cyanguwa) [inline tests/jax/test_layer.py:264]: 1. are attrs22 and attrs23 the same? I assume if window size is not set, then it is None?
2. similarly, are attrs25 and attrs28 equivalent? 
3. could we add tests for `padding_causal`?
4. in the future, could we add `causal_bottom_right` and `padding_causal_bottom_right` as well? They should be supported at the moment, but it seems that they are missing from the docstring/support matrix and the unit tests.

File: transformer_engine/jax/flax/transformer.py [hunk 2]
```diff
@@ -447,6 +447,14 @@ class DotProductAttention(nn.Module):  # pylint: disable=too-few-public-methods
 
         .. note:: THD format only supports 'padding' or 'causal_padding' mask type.
 
+       attn_mask_type       mask/sequence_descriptor       SW         softmax type
```
Comment (cyanguwa) [inline transformer_engine/jax/flax/transformer.py:450]: I guess you meant "SWA" here instead of "SW".

**Review by cyanguwa — APPROVED** [review]

File: transformer_engine/jax/flax/transformer.py (hunk 2 above)
Comment (KshitijLakhani) [reply transformer_engine/jax/flax/transformer.py:450]: @cyanguwa I'm using the convention used in AND, OR logic etc for [Don't Care](https://en.wikipedia.org/wiki/Don%27t-care_term). It suggests that does not matter whether SWA is Yes, No, None the Softmax type will be the same.
Do you suggest we be explicit and I replace X with Yes/No/None instead ?

Comment (KshitijLakhani) [reply transformer_engine/jax/flax/transformer.py]: @cyanguwa Resolved in SHA 47e151a8f40cc02ec136d4a72e4c9c090205bb30
